sqlgen/implicitnull_test.go

```diff
@@ -0,0 +1,238 @@
+	require.NoError(t, err)
+	assert.Equal(t, 0, count, "Expected null_byte to not be null")
+
+	count, err = GetCount(db.Conn, "SELECT COUNT(*) FROM implicitnull WHERE null_time IS NULL")
+	require.NoError(t, err)
+	assert.Equal(t, 0, count, "Expected null_time to not be null")
+
+	newRow.Id = 1
+
+	var rows []*ImplicitNull
+	require.NoError(t, db.Query(context.Background(), &rows, nil, nil))
+	assert.Equal(t, []*ImplicitNull{
+		newRow,
+	}, rows)
+}
+
+func GetCount(db *sql.DB, countQuery string) (int, error) {
+	dbrows, err := db.Query(countQuery)
+	if err != nil {
+		return 0, err
+	}
+
+	var c int
+	for dbrows.Next() {
+		if err := dbrows.Scan(&c); err != nil {
+			return 0, err
+		}
+	}
+
+	return c, nil
+}
+
+func TestInvalidImplicitNullField(t *testing.T) {
+	testDb, err := testfixtures.NewTestDatabase()
+	require.NoError(t, err)
+	defer testDb.Close()
+
+	type InvalidImplicitNull struct {
+		Id      int64   `sql:",primary"`
+		NullStr *string `sql:",implicitnull"` // Should not allow pointers
+	}
+
+	_, err = testDb.Exec(`
+		CREATE TABLE invalidimplicitnull (
+			id   BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,
+			null_str VARCHAR(255)
+		)
+	`)
+	require.NoError(t, err)
+
+	schema := NewSchema()
+	err = schema.RegisterType("invalidimplicitnull", AutoIncrement, InvalidImplicitNull{})
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "column null_str cannot use `implicitnull` with pointer type")
+}
+
+func TestImplicitNullFiltersWorkForNonNull(t *testing.T) {
+	tdb, db, err := setupImplicitNull()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer tdb.Close()
+
+	timeField := time.Date(2000, 10, 3, 0, 0, 0, 0, time.UTC)
+	newRow := &ImplicitNull{
+		NullStr:   "hello",
+		NullInt:   123,
+		NullFloat: 1.23,
+		NullBool:  true,
+		NullByte:  []byte("hello there"),
+		NullTime:  timeField,
+	}
+	if _, err := db.InsertRow(context.Background(), newRow); err != nil {
+		t.Error(err)
+	}
+
+	testCtx := []struct {
+		ctx context.Context
+	}{
+		{batch.WithBatching(context.Background())},
+		{context.Background()},
+	}
+
+	for _, tt := range testCtx {
+		t.Run("", func(t *testing.T) {
+			ctx := tt.ctx
+			var imp *ImplicitNull
+
+			require.NoError(t, db.QueryRow(ctx, &imp, Filter{"id": int64(1)}, nil))
+			require.NoError(t, db.QueryRow(ctx, &imp, Filter{"null_str": "hello"}, nil))
+			require.NoError(t, db.QueryRow(ctx, &imp, Filter{"null_int": int64(123)}, nil))
+			require.NoError(t, db.QueryRow(ctx, &imp, Filter{"null_float": 1.23}, nil))
+			require.NoError(t, db.QueryRow(ctx, &imp, Filter{"null_bool": true}, nil))
+			require.NoError(t, db.QueryRow(ctx, &imp, Filter{"null_byte": []byte("hello there")}, nil))
+			require.NoError(t, db.QueryRow(ctx, &imp, Filter{"null_time": timeField}, nil))
+
+			newRow.Id = 1
+			var rows []*ImplicitNull
+			require.NoError(t, db.Query(context.Background(), &rows, nil, nil))
+			assert.Equal(t, []*ImplicitNull{
+				newRow,
+			}, rows)
+		})
```

Comment:
could we also add a test for the reading that a `null` in the db reads out as the zero value?

Reply:
👍